pkg/reconciler/topology/partitionset_reconcile.go

```diff
@@ -0,0 +1,242 @@
+	}
+	for _, oldPartition := range oldPartitions {
+		// MatchExpressions need to be the same
+		oldMatchExpressions := []metav1.LabelSelectorRequirement{}
+		if oldPartition.Spec.Selector != nil {
+			oldMatchExpressions = oldPartition.Spec.Selector.MatchExpressions
+		}
+		if !equality.Semantic.DeepEqual(oldMatchExpressions, newMatchExpressions) {
+			if err := c.deletePartition(ctx, logicalcluster.From(oldPartition).Path(), oldPartition.Name); err != nil && !apierrors.IsNotFound(err) {
+				conditions.MarkFalse(
+					partitionSet,
+					topologyv1alpha1.PartitionsReady,
+					topologyv1alpha1.ErrorGeneratingPartitionsReason,
+					conditionsv1alpha1.ConditionSeverityError,
+					"old partition could not get deleted",
+				)
+				return err
+			}
+			continue
+		}
+
+		// MatchLabels need to be the same
+		partitionKey := ""
+		oldMatchLabels := map[string]string{}
+		if oldPartition.Spec.Selector != nil {
+			oldMatchLabels = oldPartition.Spec.Selector.MatchLabels
+		}
+
+		// Sorting the keys of the old partition for consistent comparison
+		sortedOldKeys := make([]string, len(oldMatchLabels))
+
+		i := 0
+		for k := range oldMatchLabels {
+			sortedOldKeys[i] = k
+			i++
+		}
+		sort.Strings(sortedOldKeys)
+		for _, key := range sortedOldKeys {
+			partitionKey = partitionKey + "+" + key + "=" + oldMatchLabels[key]
+		}
+		if _, ok := matchLabelsMap[partitionKey]; ok {
+			existingMatches[partitionKey] = placeHolder
+		} else {
+			if err := c.deletePartition(ctx, logicalcluster.From(oldPartition).Path(), oldPartition.Name); err != nil {
+				conditions.MarkFalse(
+					partitionSet,
+					topologyv1alpha1.PartitionsReady,
+					topologyv1alpha1.ErrorGeneratingPartitionsReason,
+					conditionsv1alpha1.ConditionSeverityError,
+					"old partition could not get deleted",
+				)
+				return err
+			}
+		}
+	}
+	// Create partitions when no existing partition for the set has the same selector.
+	for key, matchLabels := range matchLabelsMap {
+		if _, ok := existingMatches[key]; !ok {
+			partition := generatePartition(partitionSet.Name, newMatchExpressions, matchLabels, partitionSet.Spec.Dimensions)
+			partition.OwnerReferences = []metav1.OwnerReference{
+				*metav1.NewControllerRef(partitionSet, topologyv1alpha1.SchemeGroupVersion.WithKind("PartitionSet")),
+			}
+			_, err = c.createPartition(ctx, logicalcluster.From(partitionSet).Path(), partition)
+			if err != nil && !apierrors.IsAlreadyExists(err) {
+				conditions.MarkFalse(
+					partitionSet,
+					topologyv1alpha1.PartitionsReady,
+					topologyv1alpha1.ErrorGeneratingPartitionsReason,
+					conditionsv1alpha1.ConditionSeverityError,
+					"partition could not get created",
+				)
+				return err
+			}
+		}
+	}
+	conditions.MarkTrue(partitionSet, topologyv1alpha1.PartitionsReady)
+	return nil
+}
+
+// partition populates shard label selectors according to dimensions.
+// It only keeps selectors that have at least one match.
```

Comment:
with https://github.com/kcp-dev/kcp/pull/2513/files#r1091643038 this comment is not true, correct?

Reply:
I will make the comment clearer, what I meant: "It only keeps selectors that have at least one Shard matching them so that Partition not referring to any Shards should not get created"